sn_nat_detection/src/main.rs

```diff
@@ -0,0 +1,400 @@
+// Copyright 2024 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use clap::Parser;
+use futures::StreamExt;
+use libp2p::autonat::NatStatus;
+use libp2p::core::{multiaddr::Protocol, Multiaddr};
+use libp2p::swarm::SwarmEvent;
+use libp2p::{noise, tcp, yamux};
+use std::collections::HashSet;
+use std::error::Error;
+use std::net::Ipv4Addr;
+use std::time::Duration;
+use tracing::{debug, info, warn};
+use tracing_log::AsTrace;
+use tracing_subscriber::layer::SubscriberExt;
+use tracing_subscriber::util::SubscriberInitExt;
+use tracing_subscriber::EnvFilter;
+
+use behaviour::{Behaviour, BehaviourEvent};
+
+mod behaviour;
+
+const CONFIDENCE_MAX: usize = 2;
+const RETRY_INTERVAL: Duration = Duration::from_secs(10);
+
+/// A tool to detect NAT status of the machine. It can be run in server mode or client mode.
+/// The program will exit with an error code if NAT status is determined to be private.
+#[derive(Debug, Parser)]
+#[clap(name = "libp2p autonat")]
+struct Opt {
+    /// Port to listen on.
+    ///
+    /// `0` causes the OS to assign a random available port.
+    #[clap(long, short, default_value_t = 0)]
+    port: u16,
+
+    /// Servers to send dial-back requests to, in a 'multiaddr' format.
+    ///
+    /// A multiaddr looks like `/ip4/1.2.3.4/tcp/1200/tcp` where `1.2.3.4` is the IP and `1200` is the port.
+    /// Alternatively, the address can be written as `1.2.3.4:1200`.
+    ///
+    /// This argument can be provided multiple times to connect to multiple peers.
+    #[clap(name = "SERVER", value_name = "multiaddr", value_delimiter = ',', value_parser = parse_peer_addr)]
+    server_addr: Vec<Multiaddr>,
```

Comment:
Can we also specify that if we do not provide this argument, we are acting as a server?